HTTP 400 Errors when installing Fabric loader due to ANSI/control characters in constructed API URLs

### Bug description

When installing a Fabric loader (e.g. via `install-fabric-loader`), mc-image-helper constructs a URL that includes encoded ANSI/control characters, resulting in malformed API requests and HTTP 400 errors. Instead of a clean Minecraft version (e.g. "26.1"), parts of the URL include ANSI escape sequences (e.g. `%1B%5B39m…`) indicating that uncleaned output was used in URL construction.

This happens reliably when certain resolver outputs or log strings are passed directly into API request parameters without sanitization.

### Logs / Error Output

Example observed in container logs:

```md
[init] 2026-04-09 12:05:23-04:00 Resolved Minecraft version [mc-image-helper] 12:05:22.111 INFO  : Excluding 1 optional project(s) from Minecraft version resolution
26.1 from Modrinth projects
[init] 2026-04-09 12:05:23-04:00 Starting RCON commands
[init] 2026-04-09 12:05:23-04:00 Rcon cmds functionality enabled
[init] 2026-04-09 12:05:23-04:00 Resolving type given FABRIC
[mc-image-helper] 12:05:25.197 ERROR : 'install-fabric-loader' command failed. Version is 1.56.0: FailedRequestException: HTTP request of https://meta.fabricmc.net/v2/versions/loader/%1B%5B39m%5Bmc-image-helper%5D+12:05:22.111+INFO++:+Excluding+1+optional+project(s)+from+Minecraft+version+resolution%0A%1B%5B0;39m26.1 failed with 400 Bad Request: Fetching object content
[init] 2026-04-09 12:05:25-04:00 [ERROR] Failed to installFabric launcher given [mc-image-helper] 12:05:22.111 INFO  : Excluding 1 optional project(s) from Minecraft version resolution
26.1, LATEST, LATEST
```

The `%1B%5B39m` visible in the URL corresponds to ANSI escape codes from colored output.

### Reproduction

Steps to reproduce:
1. Use a configuration that resolves a Fabric loader (e.g. via docker-minecraft-server with `TYPE=FABRIC VERSION=26.1 ...`)
2. Run mc-image-helper with options that resolve loader versions
3. Observe that the constructed API call URL contains invalid encoded fragments

### Expected behavior

The version string used in URL construction should be cleaned to remove any ANSI/color sequences or extraneous characters so that API requests to Fabric meta endpoints are valid (e.g., `https://meta.fabricmc.net/v2/versions/loader/26.1/...`).

### Proposed fix

Before constructing API URLs (e.g. for Fabric or Modrinth), sanitize any version or slug strings by stripping ANSI escape sequences and trimming whitespace. Example in Java:

```java
private static String sanitize(String value) {
    if (value == null) return null;
    return value.replaceAll("\\u001B\\[[;\\d]*m", "").trim();
}
```

Apply this sanitization to version strings used in URL construction.

### Additional context

This has been encountered using mc-image-helper 1.56.0, and leads to invalid API calls when resolving loader or mod versions when using optionnal projects (`?`).

### Environment info

```
mc-image-helper version: 1.56.0
Minecraft loader version range: `VERSION_FROM_MODRINTH_PROJECTS: true`
Docker image: itzg/minecraft-server:latest
Context where issue observed: Docker Compose with TYPE=FABRIC and multiple Modrinth projects
```

--- a/src/main/java/me/itzg/helpers/fabric/FabricMetaClient.java
+++ b/src/main/java/me/itzg/helpers/fabric/FabricMetaClient.java
@@ -17,6 +17,7 @@ import me.itzg.helpers.files.IoStreams;
 import me.itzg.helpers.http.FileDownloadStatusHandler;
 import me.itzg.helpers.http.SharedFetch;
 import me.itzg.helpers.http.UriBuilder;
+import me.itzg.helpers.versions.McVersioning;
 import org.apache.commons.codec.binary.Hex;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -87,11 +88,11 @@ public class FabricMetaClient {
     }
 
     private static boolean isSnapshot(@Nullable String version) {
-        return version != null && version.equalsIgnoreCase("snapshot");
+        return version != null && version.equalsIgnoreCase(McVersioning.SNAPSHOT);
     }
 
     private static boolean isLatest(@Nullable String version) {
-        return version == null || version.equalsIgnoreCase("latest");
+        return version == null || version.equalsIgnoreCase(McVersioning.LATEST);
     }
 
     public Mono<String> resolveLoaderVersion(String minecraftVersion, String loaderVersion) {

--- a/src/main/java/me/itzg/helpers/fabric/FabricMetaClient.java
+++ b/src/main/java/me/itzg/helpers/fabric/FabricMetaClient.java
@@ -1,5 +1,8 @@
 package me.itzg.helpers.fabric;
 
+import static me.itzg.helpers.versions.McVersioning.isLatest;
+import static me.itzg.helpers.versions.McVersioning.isSnapshot;
+
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.ByteBuffer;
@@ -17,7 +20,6 @@ import me.itzg.helpers.files.IoStreams;
 import me.itzg.helpers.http.FileDownloadStatusHandler;
 import me.itzg.helpers.http.SharedFetch;
 import me.itzg.helpers.http.UriBuilder;
-import me.itzg.helpers.versions.McVersioning;
 import org.apache.commons.codec.binary.Hex;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -85,14 +87,6 @@ public class FabricMetaClient {
             })
             .retryWhen(Retry.backoff(retryMaxAttempts, retryMinBackoff).filter(IOException.class::isInstance))
             .checkpoint();
-    }
-
-    private static boolean isSnapshot(@Nullable String version) {
-        return version != null && version.equalsIgnoreCase(McVersioning.SNAPSHOT);
-    }
-
-    private static boolean isLatest(@Nullable String version) {
-        return version == null || version.equalsIgnoreCase(McVersioning.LATEST);
     }
 
     public Mono<String> resolveLoaderVersion(String minecraftVersion, String loaderVersion) {

--- a/src/main/java/me/itzg/helpers/fabric/InstallFabricLoaderCommand.java
+++ b/src/main/java/me/itzg/helpers/fabric/InstallFabricLoaderCommand.java
@@ -58,7 +58,7 @@ public class InstallFabricLoaderCommand implements Callable<Integer> {
 
     static class VersionOptions {
         @Option(names = "--minecraft-version", defaultValue = "latest", paramLabel = "VERSION")
-            public void setMinecraftVersion(String minecraftVersion) {
+        public void setMinecraftVersion(String minecraftVersion) {
             this.minecraftVersion = validateMinecraftVersion(minecraftVersion);
         }
         String minecraftVersion;

--- a/src/main/java/me/itzg/helpers/fabric/InstallFabricLoaderCommand.java
+++ b/src/main/java/me/itzg/helpers/fabric/InstallFabricLoaderCommand.java
@@ -1,5 +1,7 @@
 package me.itzg.helpers.fabric;
 
+import static me.itzg.helpers.versions.McVersioning.validateMinecraftVersion;
+
 import java.net.URI;
 import java.nio.file.Path;
 import java.util.concurrent.Callable;
@@ -19,6 +21,7 @@ import picocli.CommandLine.Option;
 @Slf4j
 public class InstallFabricLoaderCommand implements Callable<Integer> {
 
+    @SuppressWarnings("unused") // picocli uses it
     @Option(names = {"--help","-h"}, usageHelp = true)
     boolean help;
 
@@ -55,6 +58,9 @@ public class InstallFabricLoaderCommand implements Callable<Integer> {
 
     static class VersionOptions {
         @Option(names = "--minecraft-version", defaultValue = "latest", paramLabel = "VERSION")
+            public void setMinecraftVersion(String minecraftVersion) {
+            this.minecraftVersion = validateMinecraftVersion(minecraftVersion);
+        }
         String minecraftVersion;
 
         @Option(names = "--installer-version", paramLabel = "VERSION",

--- a/src/main/java/me/itzg/helpers/modrinth/VersionFromModrinthProjectsCommand.java
+++ b/src/main/java/me/itzg/helpers/modrinth/VersionFromModrinthProjectsCommand.java
@@ -4,18 +4,14 @@ import static me.itzg.helpers.McImageHelper.SPLIT_COMMA_NL;
 import static me.itzg.helpers.McImageHelper.SPLIT_SYNOPSIS_COMMA_NL;
 
 import java.util.Arrays;
-import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
-import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.Callable;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
 import me.itzg.helpers.errors.GenericException;
 import me.itzg.helpers.http.SharedFetchArgs;
-import me.itzg.helpers.modrinth.model.Project;
-import me.itzg.helpers.modrinth.model.Version;
 import me.itzg.helpers.modrinth.model.VersionType;
 import picocli.CommandLine.ArgGroup;
 import picocli.CommandLine.Command;
@@ -90,12 +86,14 @@ public class VersionFromModrinthProjectsCommand implements Callable<Integer> {
         // If all projects are optional, fall back to using all of them.
         final List<ProjectRef> effectiveRefs;
         if (requiredRefs.isEmpty()) {
+            // warning logs go to stderr -- won't pollute stdout
             log.warn("All projects are marked as optional — using all for version resolution");
             effectiveRefs = allRefs;
         } else {
             if (requiredRefs.size() < allRefs.size()) {
                 final long optionalCount = allRefs.size() - requiredRefs.size();
-                log.info("Excluding {} optional project(s) from Minecraft version resolution", optionalCount);
+                // debug logs go to stderr -- won't pollute stdout
+                log.debug("Excluding {} optional project(s) from Minecraft version resolution", optionalCount);
             }
             effectiveRefs = requiredRefs;
         }

--- a/src/main/java/me/itzg/helpers/versions/McVersioning.java
+++ b/src/main/java/me/itzg/helpers/versions/McVersioning.java
@@ -4,6 +4,7 @@ import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import lombok.experimental.UtilityClass;
 import me.itzg.helpers.errors.InvalidParameterException;
+import org.jspecify.annotations.Nullable;
 
 @UtilityClass
 public class McVersioning {
@@ -12,15 +13,33 @@ public class McVersioning {
     public static final String LATEST = "latest";
     public static final Pattern MC_VERSION_REGEX = Pattern.compile(
         LATEST + "|" + SNAPSHOT + "|"
-            + "(?:[0-9]{2}w[0-9]{2}[a-z]|(?:[0-9]+\\.)+[0-9]+(?:-(?:pre|rc)[0-9]+)?|\\d+\\.\\d+|\\d+)",
+            + "(?:\\d+\\.)+\\d+(?:-(?:snapshot|rc|pre)-?\\d+| Pre-Release \\d+)?"
+            + "|[0-9]{2}w[0-9]{2}[a-z_]+"
+            + "|[abc](?:\\d+\\.)+\\d+[abc]?(?:_\\d+[abc]?)?"
+            + "|(?:inf|rd)-\\d+"
+            + "|3D Shareware v1.34",
         Pattern.CASE_INSENSITIVE
     );
 
-    public static String validateMinecraftVersion(String version) {
+    public static String validateMinecraftVersion(@Nullable String version) {
+        if (version == null) {
+            return null;
+        }
+        if (version.isBlank()) {
+            return null;
+        }
         final Matcher m = MC_VERSION_REGEX.matcher(version);
         if (m.matches()) {
             return version;
         }
         throw new InvalidParameterException(String.format("%s is not a valid minecraft version", version));
+    }
+
+    public static boolean isSnapshot(@Nullable String version) {
+        return version != null && version.equalsIgnoreCase(SNAPSHOT);
+    }
+
+    public static boolean isLatest(@Nullable String version) {
+        return version == null || version.equalsIgnoreCase(LATEST);
     }
 }

--- a/src/main/java/me/itzg/helpers/versions/McVersioning.java
+++ b/src/main/java/me/itzg/helpers/versions/McVersioning.java
@@ -12,7 +12,7 @@ public class McVersioning {
     public static final String LATEST = "latest";
     public static final Pattern MC_VERSION_REGEX = Pattern.compile(
         LATEST + "|" + SNAPSHOT + "|"
-            + "[0-9]{2}w[0-9]{2}[a-z]|(?:[0-9]+\\\\.)+[0-9]+(?:-(?:pre|rc)[0-9]+)?|\\\\d+\\\\.\\\\d+|\\\\d+",
+            + "(?:[0-9]{2}w[0-9]{2}[a-z]|(?:[0-9]+\\.)+[0-9]+(?:-(?:pre|rc)[0-9]+)?|\\d+\\.\\d+|\\d+)",
         Pattern.CASE_INSENSITIVE
     );
 

--- a/src/main/java/me/itzg/helpers/versions/McVersioning.java
+++ b/src/main/java/me/itzg/helpers/versions/McVersioning.java
@@ -1,0 +1,26 @@
+package me.itzg.helpers.versions;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import lombok.experimental.UtilityClass;
+import me.itzg.helpers.errors.InvalidParameterException;
+
+@UtilityClass
+public class McVersioning {
+
+    public static final String SNAPSHOT = "snapshot";
+    public static final String LATEST = "latest";
+    public static final Pattern MC_VERSION_REGEX = Pattern.compile(
+        LATEST + "|" + SNAPSHOT + "|"
+            + "[0-9]{2}w[0-9]{2}[a-z]|(?:[0-9]+\\\\.)+[0-9]+(?:-(?:pre|rc)[0-9]+)?|\\\\d+\\\\.\\\\d+|\\\\d+",
+        Pattern.CASE_INSENSITIVE
+    );
+
+    public static String validateMinecraftVersion(String version) {
+        final Matcher m = MC_VERSION_REGEX.matcher(version);
+        if (m.matches()) {
+            return version;
+        }
+        throw new InvalidParameterException(String.format("%s is not a valid minecraft version", version));
+    }
+}

--- a/src/test/java/me/itzg/helpers/versions/McVersioningTest.java
+++ b/src/test/java/me/itzg/helpers/versions/McVersioningTest.java
@@ -1,0 +1,103 @@
+package me.itzg.helpers.versions;
+
+import static me.itzg.helpers.versions.McVersioning.*;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+import java.util.stream.Stream;
+import me.itzg.helpers.errors.InvalidParameterException;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class McVersioningTest {
+
+    @SuppressWarnings("ConstantValue")
+    @Test
+    void handlesNull() {
+        final String result = validateMinecraftVersion(null);
+        assertThat(result).isNull();
+
+        assertThat(isLatest(null)).isTrue();
+    }
+
+    @Test
+    void handlesSnapshot() {
+        assertThat(
+            validateMinecraftVersion("SNAPSHOT")
+        ).isEqualTo("SNAPSHOT");
+        assertThat(isSnapshot("SNAPSHOT")).isTrue();
+    }
+
+    @Test
+    void handlesLatest() {
+        assertThat(
+            validateMinecraftVersion("latest")
+        ).isEqualTo("latest");
+        assertThat(isLatest("latest")).isTrue();
+    }
+
+    @ParameterizedTest
+    @MethodSource("handlesValidVersions_args")
+    void handlesValidVersions(String version, boolean expectValid, boolean expectNull) {
+        if (expectValid) {
+            assertThat(validateMinecraftVersion(version))
+                .isEqualTo(expectNull ? null : version);
+        }
+        else {
+            assertThrows(InvalidParameterException.class,
+                () -> validateMinecraftVersion(version)
+            );
+        }
+    }
+
+    public static Stream<Arguments> handlesValidVersions_args() {
+
+        return Stream.of(
+            arguments("", true, true),
+            arguments("26.2-snapshot-2", true, false),
+            arguments("26.1.2", true, false),
+            arguments("26.1.2-rc-1", true, false),
+            arguments("26.1.1-rc-1", true, false),
+            arguments("26.1", true, false),
+            arguments("26.1-rc-3", true, false),
+            arguments("26.1-pre-3", true, false),
+            arguments("26.1-snapshot-11", true, false),
+            arguments("1.21.11-rc3", true, false),
+            arguments("1.21.11-pre5", true, false),
+            arguments("25w46a", true, false),
+            arguments("25w14craftmine", true, false),
+            arguments("1.21", true, false),
+            arguments("1.21-rc1", true, false),
+            arguments("1.21-pre4", true, false),
+            arguments("24w14potato", true, false),
+            arguments("23w13a_or_b", true, false),
+            arguments("1.19.1", true, false),
+            arguments("22w13oneblockatatime", true, false),
+            arguments("20w14infinite", true, false),
+            arguments("1.14.2 Pre-Release 4", true, false),
+            arguments("1.14", true, false),
+            arguments("1.14 Pre-Release 5", true, false),
+            arguments("3D Shareware v1.34", true, false),
+            arguments("1.1", true, false),
+            arguments("1.0", true, false),
+            arguments("b1.8.1", true, false),
+            arguments("b1.8", true, false),
+            arguments("b1.5_01", true, false),
+            arguments("a1.2.6", true, false),
+            arguments("a1.2.4_01", true, false),
+            arguments("a1.2.2b", true, false),
+            arguments("a1.2.2a", true, false),
+            arguments("inf-20100618", true, false),
+            arguments("c0.30_01c", true, false),
+            arguments("c0.0.13a", true, false),
+            arguments("c0.0.13a_03", true, false),
+            arguments("c0.0.11a", true, false),
+            arguments("rd-132211", true, false),
+            arguments("invalid", false, false),
+            arguments("[mc-image-helper] 12:05:22.111 INFO", false, false)
+        );
+    }
+}

